Linux support

Speykious opened the pull request:
- Resolves #6.

## PR status

Some bugs have been pointed out, and they have been fixed. The PR is complete once again!

- [x] implement basic ALSA audio context
- [x] implement WAV file reader and writer
- [x] revive WAV export
- [x] replace WinForms with [NativeFileDialogs.NET](https://github.com/Speykious/NativeFileDialogs.NET)
- [x] configure sample rate
- [x] make sample browser functional again
- [x] enumerate and configure audio output devices
- [x] read MP3 and FLAC
- [x] implement MIDI input
- [x] add "save changes" dialog back (probably as a WaveTracker custom dialog window)
- ~~implement slightly more advanced text input~~ (has been implemented already!)
- [x] revive `StereoBiQuadFilter`
- [x] maximize window on startup
- [x] make sure the Windows build isn't broken
  - [x] implement mmsystem audio backend
  - [x] MIDI input again?
  - [x] keep system sounds playing on custom message dialogs
  - [x] make sure embedded FFmpeg libraries are being found
- ~~make input fields actual input fields instead of spawning a dialogue?~~ (has been implemented already!)

Bugs to fix:
- [x] app icon problem with DesktopGL
- [x] visual stutter occurring on high audio buffer length
  - [x] Windows
  - [x] Linux
- [x] random segfaults while previewing sample in sample file manager
  - [x] Windows
  - [x] Linux

## Considerations

In https://github.com/squiggythings/WaveTracker/issues/6#issuecomment-2325044977, @tgpholly said that it would be more worthwhile contributing upstream to get better Linux support into NAudio. But I feel like instead it would be harder and take a longer time to implement, given how simple the audio interface can be for WaveTracker as far as I understand.

## Explanations

For the implementation of the ALSA audio context, I'd like to mention this handy article on [Minimal cross-platform graphics](https://zserge.com/posts/fenster). I'd probably refer to it again for the Windows backend reimplementation, and it'll serve the same purpose eventually for MacOS.

The audio thread is incredibly simple, it just reads from the audio provider and writes to ALSA in a loop. Though I had to tweak the latency and sample buffer size to get results that sounded ok on my machine and didn't underrun too much, so I'm not sure about that. I think that'll probably become a setting.

As a full disclosure, I maintain NativeFileDialogs.NET and it binds to a C library (which isn't mine).

Edit: since then, the decision to use FFmpeg for audio file reading has been accepted. I decided to use FFmpeg 4.4. It's quite old, but the main benefit is that there is already an existing Nuget package - [FFmpeg.Native](https://www.nuget.org/packages/FFmpeg.Native) - which automatically imports the libraries needed with the correct version. That way it's guaranteed that WaveTracker won't have any issue with loading them.

Here's a demo of sound, dialogs and WAV export working :D

https://github.com/user-attachments/assets/82da0919-6c0a-4e6a-8baf-5b9b68df5711

Speykious committed:
no WPF

Speykious committed:
no NAudio

Speykious committed:
Use native file dialogs

Speykious committed:
Fix inputs with custom dialog boxes

Speykious committed:
Fix dialog filters

Speykious committed:
Add Audio Linux engine

Speykious committed:
Make sure SetApartmentState is only called on Windows

Speykious committed:
Reimplement wav file reading/writing

Speykious committed:
Fix dialog filters again

Speykious committed:
Make sure file gets closed afterwards

squiggythings commented:
Thank you so much for your contributions. I hope my existing code isn't too messy or causing many problems!

If we could get cross platform audio output working without any of the windows-dependent NAudio libraries that would be a huge step forward. I think I might even separate out the logic that gathers and processes signals from all the channels in the AudioProvider.Read() loop into a function that just processes one sample's worth of audio and returns the resulting left and right values. Then that could just be linked into whatever code is feeding samples into the platform's audio buffer. I think that would be cleaner to separate the WaveTracker specific logic from the platform's audio API.

Right now I'm looking into modifying the custom input fields to accept text input, highlighting, arrow key cursor navigation, all the things you'd expect an input field to have; which, granted, is quite a lot. I think this will be the biggest roadblock towards getting the app to be cross platform, so I'd like to start chipping away at it now. Of course there are cross platform alternatives to winforms that could replace the pop-up dialog in an emergency, but I'd really really really like to have seamless field editing without opening up an ugly pop up. (It would also allow for searching in the sample browser!)

Speykious commented:
> Thank you so much for your contributions. I hope my existing code isn't too messy

So far I honestly expected all of this to be much harder than it actually was. To the contrary I find your code pretty readable and well-organized xd

HotoRas commented:
:eyes:

Speykious committed:
IAudioContext, audio output devices, sample rate

Speykious commented:
Audio devices! (ALSA description names are quite verbose...)
![image](https://github.com/user-attachments/assets/a3cc3937-a908-4827-83a0-9a48792a721a)

darkhog commented:
You could also just filter out actual duplicates (check if the name is the same as existing device on the list, if so skip adding it to the list). I think LMMS (open source FL Studio clone) does it that way.

Speykious committed:
Fix misconfiguration if first device

Speykious committed:
Discard output-only devices (they're duplicates)

Speykious commented:
I'm just not exactly sure how to know it's a duplicate, because both the full description and name are always different from each other. The output-only stuff is just a pattern I noticed...

<details>
<summary>Full print of my PCM devices</summary>

The format goes:
```
[I/O] name
  description
```
The description can span several lines, and `I` and `O` are absent if the device doesn't do input/output respectively.

```
[I/O] null
  Discard all samples (playback) or generate zero samples (capture)

[I/O] lavrate
  Rate Converter Plugin Using Libav/FFmpeg Library

[I/O] samplerate
  Rate Converter Plugin Using Samplerate Library

[I/O] speexrate
  Rate Converter Plugin Using Speex Resampler

[I/O] jack
  JACK Audio Connection Kit

[I/O] oss
  Open Sound System

[I/O] pipewire
  PipeWire Sound Server

[I/O] pulse
  PulseAudio Sound Server

[I/O] speex
  Plugin using Speex DSP (resample, agc, denoise, echo, dereverb)

[I/O] upmix
  Plugin for channel upmix (4,6,8)

[I/O] vdownmix
  Plugin for channel downmix (stereo) with a simple spacialization

[I/O] default
  Default ALSA Output (currently PipeWire Media Server)

[ /O] hdmi:CARD=HDMI,DEV=0
  HDA ATI HDMI, HDMI 0
  HDMI Audio Output

[I/O] usbstream:CARD=HDMI
  HDA ATI HDMI
  USB Stream Output

[ /O] hdmi:CARD=Generic,DEV=0
  HD-Audio Generic, HDMI 0
  HDMI Audio Output

[I/O] usbstream:CARD=Generic
  HD-Audio Generic
  USB Stream Output

[I/O] sysdefault:CARD=Generic_1
  HD-Audio Generic, ALC294 Analog
  Default Audio Device

[I/O] front:CARD=Generic_1,DEV=0
  HD-Audio Generic, ALC294 Analog
  Front output / input

[ /O] surround21:CARD=Generic_1,DEV=0
  HD-Audio Generic, ALC294 Analog
  2.1 Surround output to Front and Subwoofer speakers

[ /O] surround40:CARD=Generic_1,DEV=0
  HD-Audio Generic, ALC294 Analog
  4.0 Surround output to Front and Rear speakers

[ /O] surround41:CARD=Generic_1,DEV=0
  HD-Audio Generic, ALC294 Analog
  4.1 Surround output to Front, Rear and Subwoofer speakers

[ /O] surround50:CARD=Generic_1,DEV=0
  HD-Audio Generic, ALC294 Analog
  5.0 Surround output to Front, Center and Rear speakers

[ /O] surround51:CARD=Generic_1,DEV=0
  HD-Audio Generic, ALC294 Analog
  5.1 Surround output to Front, Center, Rear and Subwoofer speakers

[ /O] surround71:CARD=Generic_1,DEV=0
  HD-Audio Generic, ALC294 Analog
  7.1 Surround output to Front, Center, Side, Rear and Woofer speakers

[I/O] usbstream:CARD=Generic_1
  HD-Audio Generic
  USB Stream Output

[I/O] sysdefault:CARD=Microphones
  Blue Microphones, USB Audio
  Default Audio Device

[I/O] front:CARD=Microphones,DEV=0
  Blue Microphones, USB Audio
  Front output / input

[ /O] surround21:CARD=Microphones,DEV=0
  Blue Microphones, USB Audio
  2.1 Surround output to Front and Subwoofer speakers

[ /O] surround40:CARD=Microphones,DEV=0
  Blue Microphones, USB Audio
  4.0 Surround output to Front and Rear speakers

[ /O] surround41:CARD=Microphones,DEV=0
  Blue Microphones, USB Audio
  4.1 Surround output to Front, Rear and Subwoofer speakers

[ /O] surround50:CARD=Microphones,DEV=0
  Blue Microphones, USB Audio
  5.0 Surround output to Front, Center and Rear speakers

[ /O] surround51:CARD=Microphones,DEV=0
  Blue Microphones, USB Audio
  5.1 Surround output to Front, Center, Rear and Subwoofer speakers

[ /O] surround71:CARD=Microphones,DEV=0
  Blue Microphones, USB Audio
  7.1 Surround output to Front, Center, Side, Rear and Woofer speakers

[ /O] iec958:CARD=Microphones,DEV=0
  Blue Microphones, USB Audio
  IEC958 (S/PDIF) Digital Audio Output

[I/O] usbstream:CARD=Microphones
  Blue Microphones
  USB Stream Output
```

</details>

Speykious committed:
Discard sysdefault devices (they're also duplicates)

Speykious committed:
Revive StereoBiQuadFilter

Speykious commented:
So... I realized that NAudio.Core actually contains the DirectSound output implementations that link to Windows libs. In the future I might contribute to NAudio to at least make sure the Windows-specific classes get a `[SupportedOSPlatform("Windows")]`, but I'm not really confident in importing NAudio as sit stands so for now I'm copying classes over when I can. I don't think WaveTracker uses many features of NAudio anyway so it shouldn't end up being too many classes.

Speykious committed:
Revive LoopStream

Speykious committed:
Flatten audio provider (it's just a function now)

Speykious committed:
Make the sample browser functional again

Speykious commented:
The sample browser is now functional again but I *really* don't like the code I've written for that to happen... I'll see if it's still bearable when I add MP3 and FLAC parsing somehow, but even then I'd love feedback and ideas on that when I'm done.

Speykious committed:
Replce remaining `previewOut.Stop()` calls

Speykious committed:
Vendor NAudio.Midi

As it stands, depending on NAudio is not favorable because it has a strong dependency on Windows. Since adding cross-platform support to NAudio will likely take much more time than implementing what's necessary for WaveTracker, vendoring NAudio.Midi is favorable as long as NAudio remains Windows-specific.

Speykious committed:
Rename MidiInterop to Winmm

coderspark commented:
Is there a build available yet for the linux version?

Speykious commented:
You can checkout my branch and run it with `dotnet run`, but as you can see it's not complete. The tasks that are left to be done are laid out in the description of this PR.

squiggythings commented:
Progress is being made on custom text boxes. still pretty barebones and missing a lot but it's a start!
![textbox editing](https://github.com/user-attachments/assets/483e24f5-0275-4307-a941-fb3cad341c5b)

0hStormy commented:
> Progress is being made on custom text boxes. still pretty barebones and missing a lot but it's a start!

Looks cool, maybe there should be an option to choose native dialogs and custom dialogs

rcarmo commented:
I agree that the inputs and dialogs should be native--they bring input method and internationalization support for free, which custom ones will take forever to do well...

Speykious commented:
Monogame already supports IME inputs. I tested that with my Japanese IME and it already gets the characters correctly. It's just that the font used doesn't support displaying them. So I think it's just a font problem.

Speykious commented:
Update: I figured out how to get midi input on Linux and how to test it!

![image](https://github.com/user-attachments/assets/afa142cc-d9cf-4d2a-b794-b9d42b240bac)

Speykious committed:
Add SupportedOSPlatform

Speykious committed:
Merge branch 'main' into potential-linux-build

Speykious committed:
Move WaveTracker csproj in its own subfolder

It was failing to build because of duplicate assembly info.
Turns out that's because WaveTracker.Midi was there. :(

Speykious commented:
I copied `NAudio.Midi` into a separate `WaveTracker.Midi` csproj. What I'll do is implement `MidiIn` for Linux in there and let WaveTracker depend on that.

The ideal scenario is NAudio just supporting Linux out of the box, but as I said adding Linux support to NAudio is gonna take more time than doing this. I should probably keep in mind to make a PR to NAudio after all of this, at least for the MIDI part

Speykious committed:
Set audio latency from config

Speykious committed:
Better organize audio classes

Speykious committed:
Implement ALSA Midi input, remove MidiOut

Speykious commented:
Midi input implemented! Windows should still work but I didn't test that. I'll need to revive the Windows audio backend before I do that anyway.

https://github.com/user-attachments/assets/c3479b62-5f5b-4cf3-91e9-14eefe3489b8

Speykious commented:
There's bugs where the note stays on, but I have no clue if it's a WaveTracker issue or an issue with how I send note events from Alsa...

Speykious committed:
Merge branch 'main' into potential-linux-build

Speykious committed:
Keep Windows system sounds playing

squiggythings commented:
First, I can't thank you enough for your contributions. This is my first time ever making any kind of desktop application like this so contributions towards other platforms is greatly appreciated especially since I have even less experience with Linux and Mac development, and would take much much longer if I were doing it myself.

It sounds to me like the note off is only sounding when the note that was released is the same that started? i.e. Holding down C, then D, then releasing C, then releasing D would keep C playing? It could be a wavetracker issue, but it wouldn't be my first guess since the PianoInput ReceiveMidiNoteOn and ReceiveMidiNoteOff should handle this (assuming the proper midi note off events are actually being received)
It might be worth mentioning I ran into a similar issue with the original MIDI input that some MIDI controllers send note offs by sending a Note-On-but-with-velocity-0, while others actually sent a dedicated note off event followed by a separate CC value for release velocity, so I had to account for both situations. Maybe this is related? I don't know.

In other news, built-in field entry for numberboxes and textboxes is pretty much complete. Very close to being able to say goodbye to Windows Forms! You can select, copy and paste, rendering also culls if the typed text goes beyond the bounds of the box etc... It uses Monogame's built-in text input which takes IME input and all that. Although yes, the default font doesn't support every character right now, (though I expanded it a little bit to include cyrillic characters and more accented latin characters)
Just today I made the inputfields compatible with the high resolution font mode as well, which does include most characters.
![renaming instruments](https://github.com/user-attachments/assets/79e7866a-dd21-441c-a66b-686d2487f7ca)

I believe this means the only remaining windows forms dependency is the one needed to maximize the window on startup, and the one to catch the app closing event to issue an unsaved changes warning.

bard-synth commented:
this could be potentially awesome I'm looking for a new tracker to use since most of the options on Linux are rather unappealing and Renoise obviously isn't FOSS if I can potentially help with this at all I would love to

bard-synth commented:
also i don't know if this is something your planning to do but utilizing Pipewire and or JACK for the audio and midi back-end could be very powerful for routing between  applications and further supporting various modular workflows

Speykious commented:
> also i don't know if this is something your planning to do but utilizing Pipewire and or JACK for the audio and midi back-end could be very powerful for routing between  applications and further supporting various modular workflows

I'm using ALSA to have audio output on Linux (and also for MIDI input). I must admit it's my first time dealing with audio in such a manner so I haven't tried using Pipewire or JACK directly, but both of these APIs seem to be listed as audio output devices in ALSA anyway, so I think you won't have any trouble in that regard.

Speykious commented:
> I believe this means the only remaining windows forms dependency is the one needed to maximize the window on startup, and the one to catch the app closing event to issue an unsaved changes warning.

Ah, I didn't catch that, added it to the checklist, thanks.

It kinda baffles me, Monogame doesn't seem to provide anything to do this from within the game... So far the only thing I could find was [this forum question](https://community.monogame.net/t/start-in-maximized-window/12264) which solves it by manually binding to the relevant SDL function. Now that I look at the GitHub issues, [same thing but for them it doesn't work](https://github.com/MonoGame/MonoGame/issues/6960)...

I'll have to test this myself.

Speykious commented:
Upon testing it actually seems like the following code in the `App` constructor is enough to maximize the window...
```cs
graphics = new GraphicsDeviceManager(this) {
    PreferredBackBufferWidth = GraphicsAdapter.DefaultAdapter.CurrentDisplayMode.Width,
    PreferredBackBufferHeight = GraphicsAdapter.DefaultAdapter.CurrentDisplayMode.Height,
};
graphics.ApplyChanges();
```

Speykious committed:
Maximize window on startup

squiggythings commented:
I had tried this originally but at least on windows, it doesn't actually maximize the window, just resizes it to fit the screen (though goes past the task bar), but it's technically still in windowed mode.
![image](https://github.com/user-attachments/assets/3eaf8035-3fee-46b4-ad00-145294df1d57)
as opposed to actually maximizing the window: 
![image](https://github.com/user-attachments/assets/435a6a81-6427-48f3-b749-24e299499339)

Speykious commented:
I see, so I guess it depends on the platform...
Weirdly enough, calling `SDL_MaximizeWindow` on the window handle doesn't work at all on my Linux machine, which is weird because `SDL_MaximizeWindow` definitely works by itself in an SDL game. What happens on your side if you do both the `graphics.ApplyChanges()` code above and also call `SDL_MaximizeWindow(Window.Handle)` in `Initialize`?

Speykious commented:
Update: I just tried the combination of both on my dad's very very slow Windows 10 computer. It was extremely painful but it works :v

Speykious committed:
Call SDL_MaximizeWindow

Speykious commented:
> I don't know what the best practices are for this kind of thing but I'm wondering if it's a good idea to make a class that handles system specific stuff (like playing system sounds, getting the window to maximize/subscribe to an onclosing event) with a preprocessor conditionals for each platform.

I'm not sure. So far there's only one place where Windows system sounds are being played so I wouldn't create a class just for that. And as for the onclosing event, I'll have to see how I'm gonna revive the exit-save dialog.

In my opinion, the best, most ideal way to support every platform would be to use something other than Monogame, maybe with SDL3 for windowing as a base. One huge advantage is that it could then run natively on Mac M1, M2 etc. devices, because apparently Monogame doesn't support Arm64. :c
But that would take a long time (sure, the rendering code is all in the `Graphics` class and it wouldn't need to change much, but lots of classes use Monogame's various other things like the game loop and retrieving keyboard/mouse state) so I guess hacking around things like I'm doing for maximizing here is more viable as it stands.

Speykious committed:
Restore save dialog

Speykious committed:
Revive Sample.SaveToDisk

paiyayay commented:
WOW! this looks 90% done. Well done sir.

Speykious commented:
For the audio file formats, I'm thinking of copying the audio file reading code from NAudio again, but I'm also contemplating the option of using FFmpeg through something like [FFMediaToolkit](https://www.nuget.org/packages/FFMediaToolkit) and an embed of the FFmpeg runtime libraries... Unfortunately there isn't anything close to [Symphonia](https://github.com/pdeljanov/Symphonia) in the C# ecosystem it seems. :/

Speykious committed:
Merge branch 'main' into potential-linux-build

TheoCGaming commented:
Once this pull request is finished, does this mean that macOS support is also added?

Speykious commented:
Yeah, somewhat, though it doesn't mean that MacOS will be supported yet. The key blocker for MacOS is that Monogame doesn't support Arm, so WaveTracker currently cannot support Macs with M1/M2 processors natively... There's a tool for Mac to run x64 programs on Arm, I'll have to see if I can use that.

Speykious committed:
Use DirectX backend on Windows

Speykious committed:
AudioLinuxContext -> AudioContextLinux

TheoCGaming commented:
was curious because i want to see if i can get it on my school's macs >:3

Speykious committed:
Remove now-unused EnterText dialog

Speykious committed:
Start Windows audio backend implementation

Speykious commented:
I just realized that if I want to use `SDL_MaximizeWindow` to maximize the window, I can't make Windows use Monogame's DirectX backend, because that one doesn't use SDL under the hood. So either I have to make WaveTracker use DesktopGL on every platform, or I somehow make it so that on Windows it binds the window to a Windows form and maximizes that like it was doing before... :/

Speykious commented:
My Windows VM was so slow that it was starting to be really painful to test anything, and my dad's Windows laptop was no better. So I said fuck it and reset my machine to have a dual boot. I can finally be productive again
![image](https://github.com/user-attachments/assets/2c7551bd-6126-4c94-b28f-3f9339c04651)

Speykious committed:
Set window position to zero

Speykious committed:
Use DesktopGL on all platforms

The WindowsDX backend doesn't use SDL, which makes it less convenient to maximize the window.

Speykious committed:
Implement audio output on Windows

Speykious committed:
Merge branch 'main' into potential-linux-build

Speykious committed:
Fix Windows autio output; implement sample rate and latency change

Speykious committed:
Safety: make sure buffer has a minimum length on Windows

Speykious commented:
> For the audio file formats, I'm thinking of copying the audio file reading code from NAudio again, but I'm also contemplating the option of using FFmpeg through something like [FFMediaToolkit](https://www.nuget.org/packages/FFMediaToolkit) and an embed of the FFmpeg runtime libraries... Unfortunately there isn't anything close to [Symphonia](https://github.com/pdeljanov/Symphonia) in the C# ecosystem it seems. :/

@squiggythings I'd love your input on this. Do you think it's a good idea to go with FFmpeg here or should I just copy NAudio?

squiggythings commented:
@Speykious I think FFmpeg seems like a good idea. If I'm understanding correctly it seems like that would have much more comprehensive support for almost all the other audio formats such as .ogg (#57) ? I also wonder though, would that allow being able to play/audition/loop the files directly when navigating through the browser without needing to read them to float arrays and then send it through wavetracker's audio engine, which can be a bit slower to initially import a file and would cause a noticeable hang when skimming through a bunch of samples in the browser, especially with longer samples.

Speykious commented:
It could allow that, yeah. I'll try focusing on just implementing the simplest FFmpeg code for now though. :D

Speykious committed:
Merge branch 'main' into potential-linux-build

Speykious committed:
Merge branch 'main' into potential-linux-build

Speykious committed:
Read audio files with FFmpeg

Speykious committed:
Merge branch 'main' into potential-linux-build

Speykious commented:
On Windows, it's somehow not able to load the FFmpeg libraries even though the path is correct (I checked like 10 times)...

defensem3ch commented:
@Speykious This is the log I get when trying to debug using VSC 2022 on Windows 10:
```
'WaveTracker.exe' (CoreCLR: DefaultDomain): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Private.CoreLib.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded '.\WaveTracker\WaveTracker\bin\Debug\net8.0\WaveTracker.dll'. Symbols loaded.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Runtime.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'c:\program files\microsoft visual studio\2022\community\common7\ide\commonextensions\microsoft\hotreload\Microsoft.Extensions.DotNetDeltaApplier.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.IO.Pipes.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Linq.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Collections.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Console.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Threading.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Runtime.InteropServices.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Threading.Overlapped.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Security.AccessControl.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Security.Principal.Windows.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Security.Claims.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Runtime.Loader.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Collections.Concurrent.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded '.\WaveTracker\WaveTracker\bin\Debug\net8.0\MonoGame.Framework.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded '.\WaveTracker\WaveTracker\bin\Debug\net8.0\FFmpeg.AutoGen.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\netstandard.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.ComponentModel.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Threading.Thread.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Text.Json.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded '.\WaveTracker\WaveTracker\bin\Debug\net8.0\WaveTracker.Midi.dll'. Symbols loaded.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Memory.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Text.Encoding.Extensions.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Text.Encodings.Web.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Runtime.Intrinsics.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Numerics.Vectors.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Private.Uri.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Reflection.Emit.Lightweight.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Reflection.Primitives.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Reflection.Emit.ILGeneration.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Runtime.Serialization.Primitives.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded '.\WaveTracker\WaveTracker\bin\Debug\net8.0\protobuf-net.Core.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Exception thrown: 'System.Text.Json.JsonException' in System.Text.Json.dll
'WaveTracker.exe' (CoreCLR: clrhost): Loaded '.\WaveTracker\WaveTracker\bin\Debug\net8.0\FontStashSharp.MonoGame.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.10\System.Text.RegularExpressions.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded '.\WaveTracker\WaveTracker\bin\Debug\net8.0\FontStashSharp.Base.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded '.\WaveTracker\WaveTracker\bin\Debug\net8.0\FontStashSharp.Rasterizers.StbTrueTypeSharp.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'WaveTracker.exe' (CoreCLR: clrhost): Loaded '.\WaveTracker\WaveTracker\bin\Debug\net8.0\StbTrueTypeSharp.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Exception thrown: 'System.DllNotFoundException' in FFmpeg.AutoGen.dll
An unhandled exception of type 'System.DllNotFoundException' occurred in FFmpeg.AutoGen.dll
Unable to load DLL 'avutil.56 under .\WaveTracker\WaveTracker\bin\Debug\net8.0\runtimes\win-x64\native': The specified module could not be found.
```
In the `.\WaveTracker\WaveTracker\bin\Debug\net8.0\runtimes\win-x64\native` folder I have three files: `nfd.dll`, `SDL2.dll`, and `soft_oal.dll`. Maybe this information will be of some use?

Speykious commented:
Sorry, I forgot to push changes to the Windows path. That path in your logs is not correct since FFmpeg puts them under `win7-x64` instead. I'll push my latest changes soon.

Speykious commented:
Oh my god... The warning was right there and I ignored it until now...!
```
C:\Program Files\dotnet\sdk\8.0.403\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(284,5): warning NETSDK1206: Found version-specific or distr 
       ibution-specific runtime identifier(s): win7-x64, win7-x86. Affected libraries: FFmpeg.Native. In .NET 8.0 and higher, assets for version-specific and dis 
       tribution-specific runtime identifiers will not be found by default. See https://aka.ms/dotnet/rid-usage for details.
```

Speykious commented:
Turns out the answer was actually a huge hack that involves modifying the process's path because Windows can't do it by itself -_-

Speykious committed:
Fix Windows path for loading FFmpeg libraries

Speykious committed:
Small modifications

Speykious committed:
Fix and document the Windows audio context

It wasn't managing resources properly upon resetting for further configuration.

Speykious committed:
Remove unused imports

Speykious committed:
Remove Windows-only PreBuild task

I generated a new DesktopGL Monogame project from their template, and that as just not present, so I'll assume that it's not needed.

Speykious committed:
Add missing implementation ref links

Speykious committed:
Remove unused Wav methods

Speykious committed:
Make sure audio reader is disposed

Speykious committed:
Remove leftover print statements

Speykious commented:
I wonder if I should remove the unused MIDI events in `WaveTracker.Midi`. They probably take most of the 7k lines of code in this PR.

Speykious committed:
Remove all unused events in WaveTracker.Midi

Speykious commented:
That only removed 2k lines??? damn

knuxify commented:
The "Open themes folder..." button in Preferences -> Colors seems to be hardcoded to open explorer, and crashes under Linux:

```
Unhandled exception. System.ComponentModel.Win32Exception (2): An error occurred trying to start process 'explorer.exe' with working directory '/home/knuxify/clones/WaveTracker'. No such file or directory
   at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   at WaveTracker.UI.ConfigurationDialog.AppearancePage.Update() in /home/knuxify/clones/WaveTracker/WaveTracker/Source/UI/ConfigurationDialog.cs:line 475
   at WaveTracker.UI.ConfigurationDialog.Update() in /home/knuxify/clones/WaveTracker/WaveTracker/Source/UI/ConfigurationDialog.cs:line 298
   at WaveTracker.UI.Dialogs.Update() in /home/knuxify/clones/WaveTracker/WaveTracker/Source/UI/Dialogs.cs:line 84
   at WaveTracker.App.Update(GameTime gameTime) in /home/knuxify/clones/WaveTracker/WaveTracker/Source/App.cs:line 359
   at Microsoft.Xna.Framework.Game.DoUpdate(GameTime gameTime)
   at Microsoft.Xna.Framework.Game.Tick()
   at Microsoft.Xna.Framework.SdlGamePlatform.RunLoop()
   at WaveTracker.Program.Main(String[] args) in /home/knuxify/clones/WaveTracker/WaveTracker/Source/Program.cs:line 16
Aborted
```

Could probably be changed to use xdg-open instead(?).

Speykious commented:
Oh, thanks for the catch! That completely went past me.

Speykious committed:
Fix opening files and links externally

squiggythings commented:
Hey apologies it has taken me so long to take a look at this! I have been incredibly busy the past few weeks. A couple things I'm noticing/having trouble with (on windows):
1. I think this is a known bug with monogame using DesktopGL that it doesn't use the custom Icon.ico, even when it is set to use one it will default to the monogame logo instead of the wavetracker logo. If anyone knows of a workaround or fix for this please let me know!
2. I'm noticing some stuttering with visuals if the buffer length is set too high, especially noticeable with the visualizer. I'm probably misunderstanding something fundamental about how the new audio engine is working but it almost seems as if the drawing code is waiting for the audio buffer to be filled before it draws again whereas previously the drawing code would just access channel and playback data mid-buffer in whatever state it was at the point the draw was called.

Speykious commented:
Ok, I compared both branches today and I can definitely see the difference in visual stutter. The main branch just appears to be delayed while my PR branch appears to wait every time after a bit. I think it's gonna take a little while to figure out what's causing this.

Speykious committed:
Fix DesktopGL app icon bug

Speykious committed:
Fix indentation

Speykious committed:
Linux: Write audio in non-blocking mode

Speykious commented:
Turns out it might be simpler than I expected. It's already fixed on Linux, I just needed to set Alsa to write audio in non-blocking mode, and just wait at the beginning of the write if it happens to be busy.
I hope that it's gonna be as simple for Windows now :v

Speykious commented:
As a side note, before the minimum latency I could have before the audio bugs out was 15 ms, now it seems to be 25 ms instead.

Speykious committed:
Fix audio visualization stuttering on Windows

Speykious commented:
FINALLY I have a solution to this problem... I noticed that it just wasn't waiting any time in between the writes, so I made it wait manually. Let me know if there's any issue left. I *think* I fixed any issue related to underflow but I'm not 100% sure.

Speykious committed:
Polish Windows audio write logic

Speykious commented:
Nevermind, clipping still appears after running the app for about 1 minute... Not sure what to do about it.

Speykious committed:
Fix wrong Windows buffer latency

Speykious committed:
Ensure better audio visualization smoothness on Windows

Speykious committed:
Remove `AvailableFrames`

Speykious committed:
Try to revive autio context after exception

Speykious committed:
Windows audio context: close before opening if already opened

Speykious committed:
Explicit variable initialization

squiggythings commented:
Finally getting a chance to take a look now. Everything seems to be working okay on windows except for a crash I sometimes get when previewing samples in the sample browser. Not sure exactly what's causing it but I'm going to try to see if I can find out.

Speykious commented:
Thank you! I've noticed it as well. There's probably some low-level thing I've written in there that isn't quite sound/safe, although I didn't manage to find the problem. I'd say the best places to look are the audio engine implementations.

Speykious committed:
Fix a previous bad partial commit

Speykious commented:
Ok, it seems I haven't been careful enough with my audio engine. It's writing in the audio thread but changing the sample rate on the main thread. I'll try to think of the best way to have everything related to managing the audio output on a single thread.

Speykious committed:
minor code change

Speykious committed:
Use messages to control audio context exclusively from audio thread

Speykious commented:
I still have to test on Windows but I at least fixed the crash on Linux

Speykious committed:
Do not send command to stop audio thread if audio thread doesn't exist

Speykious committed:
Fix audio staggering on Windows

Speykious commented:
I fixed it yesterday on Windows, I think it should be good now

peppidesu commented:
Hi, what things still need to happen for this to be merged?

Astralchroma commented:
I don't think anything needs to be done right now. Just it needs to be looked over.

user0-07161 commented:
Linux user here. NativeFileDialogs doesn't work.
```
Unhandled exception. NativeFileDialogs.Net.NfdException: The name is not activatable
   at NativeFileDialogs.Net.ResultExtensions.ToNfdStatus(Result result)
   at NativeFileDialogs.Net.Nfd.OpenDialog(String& outPath, IDictionary`2 filters, String defaultPath)
   at WaveTracker.SaveLoad.<>c__DisplayClass53_0.<SetFilePathThroughOpenDialog>b__0() in /home/user0/WaveTracker/WaveTracker/Source/Tracker/SaveLoad.cs:line 477
```

Speykious commented:
Hey, I maintain NativeFileDialogs.NET, so if there's a problem I can try fixing it xd
I'll look into that bug in a few days

user0-07161 commented:
Anyway, this PR is great work, really like it

Speykious commented:
Very late update on this bug: I just can't reproduce it. If it manages to come up again in the future I'll make a future PR I guess

Speykious commented:
Thank you so much!!

Do mention me on issues if any problem with the audio engine or something along those lines resurface.

RubberDuckyDJ commented:
@Speykious You actually rock so much.
